File: src/nikeno/Tenki/service/WidgetUpdateService.java
package nikeno.Tenki.service;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.app.Service;
import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.IBinder;
import android.os.SystemClock;
import android.text.SpannableString;
import android.text.format.DateUtils;
import android.text.style.ForegroundColorSpan;
import android.util.Log;
import android.widget.RemoteViews;

import java.util.Calendar;
import java.util.Locale;

import nikeno.Tenki.Const;
import nikeno.Tenki.Downloader;
import nikeno.Tenki.MainActivity;
import nikeno.Tenki.R;
import nikeno.Tenki.TenkiWidgetProvider;
import nikeno.Tenki.YahooWeather;
import nikeno.Tenki.YahooWeather.Day;
import nikeno.Tenki.activity.TenkiWidgetConfigure;
import nikeno.Tenki.activity.TenkiWidgetConfigure.WidgetConfig;

public class WidgetUpdateService extends Service {

	static final String TAG = WidgetUpdateService.class.getSimpleName();

	UpdateTask mUpdateTask;


	public static void start(Context context) {
		context.startService(new Intent(context, WidgetUpdateService.class));
	}

	/**
	 * デフォルトの間隔でアラームを設定
	 * @param context
	 */
	public static void setAlarm(Context context) {
		setAlarm(context, 1*DateUtils.HOUR_IN_MILLIS);
	}

	/**
	 * アラームを設定
	 * @param context
	 * @param interval
	 */
	public static void setAlarm(Context context, long interval) {
		if (Const.DEBUG) Log.d(TAG, "setAlarm");

		AlarmManager am = (AlarmManager)context.getSystemService(ALARM_SERVICE);

		Intent intent = new Intent(context, WidgetUpdateService.class);
		PendingIntent pi = PendingIntent.getService(context,
				0, intent, 0);

		am.setRepeating(AlarmManager.ELAPSED_REALTIME,
				SystemClock.elapsedRealtime() + interval,
				interval, pi);
	}

	@Override
	public IBinder onBind(Intent intent) {
		return null;
	}

	@Override
	public void onStart(Intent intent, int startId) {
		if (Const.DEBUG) {
			Log.d(TAG, "UpdateService started");
		}
		super.onStart(intent, startId);

		startUpdateTask();
	}

	void onUpdateTaskFinished() {
		mUpdateTask = null;
		stopSelf();
	}

	private void startUpdateTask() {
		if (mUpdateTask == null) {
			mUpdateTask = new UpdateTask() {
				@Override
				protected void onPostExecute(Object result) {
					super.onPostExecute(result);
					onUpdateTaskFinished();
				}
				@Override
				protected void onCancelled() {
					super.onCancelled();
					onUpdateTaskFinished();
				}
			};
			mUpdateTask.execute(this);
		}
	}

	// -----------------------------------
	// UpdateTask
	// -----------------------------------

	private static class UpdateTask extends AsyncTask<Object, Integer, Object> {

		@Override
		protected Object doInBackground(Object... params) {
			Context context = (Context)params[0];
			updateWidgets(context);
			return null;
		}

		public static void updateWidgets(Context context) {
			ComponentName thisWidget = new ComponentName(context, TenkiWidgetProvider.class);
			AppWidgetManager manager = AppWidgetManager.getInstance(context);
			int [] ids = manager.getAppWidgetIds(thisWidget);

			WidgetTheme theme = new WidgetTheme(context);

			Log.d(TAG, "updateWidgets count:"+ids.length);
			for (int id: ids) {
				WidgetConfig config = TenkiWidgetConfigure.getWidgetConfig(context, id);

				try {
					byte [] html = Downloader.download(config.url, Const.HTML_SIZE_MAX,
							10*DateUtils.SECOND_IN_MILLIS, true);
					YahooWeather data = YahooWeather.parse(html);

					RemoteViews views = buildUpdate(context, id, data, theme);

					Intent i = new Intent(context, MainActivity.class);
					i.setData(Uri.parse(config.url));
					PendingIntent pi = PendingIntent.getActivity(context, 0,
							i, 0);
					views.setOnClickPendingIntent(R.id.container, pi);
					manager.updateAppWidget(id, views);
				} catch (Exception e) {
					e.printStackTrace();
				}

			}
		}

		public static RemoteViews buildUpdate(Context context, int id, YahooWeather data,
				WidgetTheme theme) {
			RemoteViews views = new RemoteViews(context.getPackageName(),
					R.layout.widget);

			String time = DateUtils.formatDateTime(context, System.currentTimeMillis(),
					DateUtils.FORMAT_SHOW_TIME);
			views.setTextViewText(R.id.title,data.areaName);
			views.setTextViewText(R.id.time, time + " 更新");

			int [] heads = new int [] {
					R.id.h0, R.id.h1, R.id.h2, R.id.h3, R.id.h4,
					R.id.h5, R.id.h6, R.id.h7
				};
			int [] cols = new int [] {
					R.id.t0, R.id.t1, R.id.t2, R.id.t3, R.id.t4,
					R.id.t5, R.id.t6, R.id.t7
				};
			int [] imgs = new int [] {
					R.id.i0, R.id.i1, R.id.i2, R.id.i3, R.id.i4,
					R.id.i5, R.id.i6, R.id.i7
				};

			final int HOUR = 60 * 60 * 1000;
			Calendar nowJapan = Calendar.getInstance(Locale.JAPAN);
			long now = nowJapan.getTime().getTime() - 3 * HOUR;
			int col = 0;

			for (int y=0; y<2; y++) {
				Day day = (y == 0 ? data.today : data.tomorrow);

				long baseTime = day.date.getTime();

				for (int x=0; x<8; x++) {
					YahooWeather.Hour h = day.hours[x];

					boolean enabled = (baseTime + h.hour * HOUR) > now;
					if (enabled && col < cols.length) {

						views.setTextViewText(heads[col],
								String.valueOf(h.hour) +"時");

						StringBuilder sb = new StringBuilder();

						sb.append(h.text + "\n");
						int textEnd = sb.length();

						sb.append(h.temp + "℃\n");
						int tempEnd = sb.length();

						//sb.append(h.humidity + "\n");
						int humidEnd = sb.length();

						sb.append(h.rain + "㍉");
						int rainEnd = sb.length();

						String imageUrl = h.getImageUrl(true);
						Log.d(TAG, "imagerUrl:"+imageUrl);
						if (imageUrl != null) {
							int bmpId = getBitmap(context, imageUrl);;
							if (bmpId != -1) {
								views.setImageViewResource(imgs[col], bmpId);
							}
							else {
								Bitmap bmp = null;
								try {
									bmp = Downloader.downloadImage(imageUrl, 8000, 0);
									views.setImageViewBitmap(imgs[col], bmp);
								} catch (Exception e) {
									e.printStackTrace();
								}
							}
						}

						SpannableString ss = new SpannableString(sb);
						ss.setSpan(new ForegroundColorSpan(theme.textColor), 0, textEnd,
								SpannableString.SPAN_INCLUSIVE_INCLUSIVE);
						ss.setSpan(new ForegroundColorSpan(theme.tempColor), textEnd, tempEnd,
								SpannableString.SPAN_INCLUSIVE_INCLUSIVE);
						ss.setSpan(new ForegroundColorSpan(theme.humidColor), tempEnd, humidEnd,
								SpannableString.SPAN_INCLUSIVE_INCLUSIVE);
						ss.setSpan(new ForegroundColorSpan(theme.rainColor), humidEnd, rainEnd,
								SpannableString.SPAN_INCLUSIVE_INCLUSIVE);
						views.setTextViewText(cols[col], ss);

						col++;
					}
				}
			}

			return views;
		}

	}

	private static final String [] bmpNames = new String [] {
		"psun", "psnow", "psleet",
		"prain", "pmoon", "pclouds"
	};
	private static final int [] bmpIds = new int [] {
		R.drawable.psun, R.drawable.psnow, R.drawable.psleet,
		R.drawable.prain, R.drawable.pmoon, R.drawable.pclouds
	};

	public static int getBitmap(Context c, String url) {
		for (int j=0; j<bmpNames.length; j++) {
			if (url.indexOf(bmpNames[j]) != -1) {
				return bmpIds[j];
			}
		}
		return -1;
	}

	public static class WidgetTheme {
		public final int timeColor;
		public final int tempColor;
		public final int rainColor;
		public final int humidColor;
		public final int textColor;

		public WidgetTheme(Context context) {
//			Resources res = context.getResources();

			timeColor = 0xffd0d0d0;
			tempColor = 0xffFFAC59;
			rainColor = 0xff8EC7FF;
			humidColor = 0xff8EFF8E;
			textColor = 0xffe0e0e0;
		}
	}
}

File: app/src/main/java/nikeno/Tenki/service/WidgetUpdateService.kt
package nikeno.Tenki.service

import android.app.IntentService
import android.app.Notification
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.net.Uri
import android.os.Build
import android.os.IBinder
import android.text.SpannableString
import android.text.format.DateUtils
import android.text.style.ForegroundColorSpan
import android.util.Log
import android.widget.RemoteViews
import kotlinx.coroutines.runBlocking
import kotlinx.datetime.Clock
import nikeno.Tenki.MainActivity
import nikeno.Tenki.R
import nikeno.Tenki.TenkiApp
import nikeno.Tenki.TenkiApp.Companion.from
import nikeno.Tenki.TenkiWidgetProvider
import nikeno.Tenki.appwidget.weatherwidget.WeatherWidgetPrefs
import nikeno.Tenki.feature.fetcher.WeatherFetcher
import nikeno.Tenki.feature.weather.YahooWeather
import nikeno.Tenki.feature.weather.YahooWeather.Hour
import nikeno.Tenki.feature.weather.YahooWeatherHtmlParser
import nikeno.Tenki.util.PendingIntentCompat
import org.koin.compose.koinInject
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import java.io.IOException
import java.net.UnknownHostException
import java.util.Calendar
import java.util.Locale
import kotlin.math.max
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.minutes


class WidgetUpdateServiceComponent : KoinComponent {
    val fetcher: WeatherFetcher by inject()
}

class WidgetUpdateService : IntentService("WidgetUpdateService") {

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onHandleIntent(intent: Intent?) {
        Log.d(TAG, "onHandleIntent $intent")

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Log.d(TAG, "startForeground")
            val n = Notification.Builder(this, TenkiApp.N_CH_WIDGET_UPDATE_SERVICE)
                .setOngoing(true)
                .build()
            startForeground(TenkiApp.N_ID_WIDGET_UPDATE_SERVICE, n)
        }

        Log.d(TAG, "ウィジェット更新中")
        val isManualUpdate = (intent != null && ACTION_MANUAL_UPDATE == intent.action)
        runBlocking {
            UpdateTask().updateWidgets(this@WidgetUpdateService, isManualUpdate)
        }
        Log.d(TAG, "ウィジェット更新完了")
    }

    internal class UpdateTask {

        val fetcher = WidgetUpdateServiceComponent().fetcher

        fun updateWidgets(context: Context, isManualUpdate: Boolean) {
            val manager = AppWidgetManager.getInstance(context)
            val ids = manager.getAppWidgetIds(
                ComponentName(context, TenkiWidgetProvider::class.java)
            )

            val theme = WidgetTheme(context)

            if (isManualUpdate) {
                for (id in ids) {
                    manager.updateAppWidget(id, createProgressView(context))
                }
            }

            Log.d(TAG, "updateWidgets count:" + ids.size)
            var error: Throwable? = null
            for (id in ids) {
                for (retryCount in 0..0) {
                    try {
//                        if (retryCount > 0) {
//                            Thread.sleep((retryCount - 1) * 3000);
//                        }
                        runBlocking { updateWidget(context, manager, id, theme, false) }
                        error = null
                        break
                    } catch (e: UnknownHostException) {
                        e.printStackTrace()
                        error = e
                    } catch (e: Exception) {
                        e.printStackTrace()
                        error = e
                        break
                    }
                }
                if (error != null) {
                    try {
                        Log.d(TAG, "キッシュで更新")
                        runBlocking { updateWidget(context, manager, id, theme, true) }
                    } catch (e: Exception) {
                        manager.updateAppWidget(id, createErrorView(context, error))
                    }
                }
            }
        }

        @Throws(Exception::class)
        suspend fun updateWidget(
            context: Context, manager: AppWidgetManager,
            id: Int, theme: WidgetTheme, forceCache: Boolean
        ) {
            val config = WeatherWidgetPrefs.getWidgetConfig(context, id)

            val weather = if (forceCache) {
                fetcher.getWeather(config.url, 0)
            } else {
                fetcher.getWeather(
                    config.url,
                    Clock.System.now().minus(15.minutes).toEpochMilliseconds()
                )
            }

            val views = when (weather) {
                is WeatherFetcher.WeatherResult.Error ->
                    createErrorView(context, weather.error)

                is WeatherFetcher.WeatherResult.Success ->
                    buildUpdate(context, id, weather.data, theme, forceCache)
            }

            val i = Intent(context, MainActivity::class.java)
            i.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_NO_HISTORY)
            i.setData(Uri.parse(config.url))
            val pi = PendingIntent.getActivity(
                context, 0,
                i, PendingIntentCompat.FLAG_MUTABLE
            )
            views.setOnClickPendingIntent(R.id.container, pi)
            manager.updateAppWidget(id, views)
        }

        fun getManualReloadPendingIntent(context: Context?): PendingIntent {
            val i = Intent(context, WidgetUpdateService::class.java)
            i.setAction(ACTION_MANUAL_UPDATE)
            return PendingIntent.getService(
                context, 0, i,
                PendingIntentCompat.FLAG_MUTABLE
            )
        }

        fun createErrorView(context: Context, e: Throwable): RemoteViews {
            val views = RemoteViews(
                context.packageName,
                R.layout.widget_error
            )
            val time = DateUtils.formatDateTime(
                context, System.currentTimeMillis(),
                DateUtils.FORMAT_SHOW_TIME or DateUtils.FORMAT_SHOW_DATE
            )
            views.setTextViewText(R.id.errorMessage, time + ":" + e.message)
            views.setOnClickPendingIntent(R.id.errorMessage, getManualReloadPendingIntent(context))
            return views
        }

        fun createProgressView(context: Context): RemoteViews {
            val views = RemoteViews(
                context.packageName,
                R.layout.widget_progress
            )
            views.setOnClickPendingIntent(R.id.progress, getManualReloadPendingIntent(context))
            return views
        }

        fun buildUpdate(
            context: Context, id: Int, data: YahooWeather,
            theme: WidgetTheme, forceCache: Boolean
        ): RemoteViews {
            val views = RemoteViews(
                context.packageName,
                R.layout.widget
            )

            val time = DateUtils.formatDateTime(
                context, System.currentTimeMillis(),
                DateUtils.FORMAT_SHOW_TIME or DateUtils.FORMAT_SHOW_DATE
            )
            views.setTextViewText(R.id.title, data.areaName)
            if (forceCache) {
                views.setTextViewText(
                    R.id.time,
                    context.getString(R.string.updateErrorTimeFmt, time)
                )
            } else {
                views.setTextViewText(R.id.time, context.getString(R.string.updateTimeFmt, time))
            }
            views.setOnClickPendingIntent(R.id.time, getManualReloadPendingIntent(context))
            views.setOnClickPendingIntent(R.id.h6, getManualReloadPendingIntent(context))
            views.setOnClickPendingIntent(R.id.h7, getManualReloadPendingIntent(context))

            val heads = intArrayOf(
                R.id.h0, R.id.h1, R.id.h2, R.id.h3, R.id.h4,
                R.id.h5, R.id.h6, R.id.h7
            )
            val cols = intArrayOf(
                R.id.t0, R.id.t1, R.id.t2, R.id.t3, R.id.t4,
                R.id.t5, R.id.t6, R.id.t7
            )
            val imgs = intArrayOf(
                R.id.i0, R.id.i1, R.id.i2, R.id.i3, R.id.i4,
                R.id.i5, R.id.i6, R.id.i7
            )

            val HOUR = 60 * 60 * 1000
            val nowJapan = Calendar.getInstance(Locale.JAPAN)
            val now = nowJapan.time.time - 3 * HOUR
            var col = 0

            for (y in 0..1) {
                val day: YahooWeather.Day = (if (y == 0) data.today else data.tomorrow)

                val baseTime: Long = day.date.toEpochMilliseconds()

                for (x in 0..7) {
                    val h: Hour = day.hours.get(x)

                    val enabled = (baseTime + h.hour * HOUR) > now
                    if (enabled && col < cols.size) {
                        views.setTextViewText(
                            heads[col],
                            h.hour.toString() + "時"
                        )

                        val sb = StringBuilder()

                        //                        sb.append(h.text).append("\n");
                        val textEnd = sb.length

                        sb.append(h.temp).append("℃\n")
                        val tempEnd = sb.length

                        //sb.append(h.humidity + "\n");
                        val humidEnd = sb.length

                        sb.append(h.rain).append("㍉")
                        val rainEnd = sb.length

                        val imageUrl = h.getImageUrl(true)
                        Log.d(TAG, "imagerUrl:$imageUrl")
                        if (imageUrl != null) {
                            val bmpId = getBitmapIndexFromUrl(context, imageUrl)
                            if (bmpId != -1) {
                                views.setImageViewResource(imgs[col], bmpId)
                            } else {
                                val bmp = getIconManager(context).getIcon(imageUrl)
                                views.setImageViewBitmap(imgs[col], bmp)
                            }
                        }

                        val ss = SpannableString(sb)
                        //                        ss.setSpan(new ForegroundColorSpan(theme.textColor), 0, textEnd,
//                                SpannableString.SPAN_INCLUSIVE_INCLUSIVE);
                        ss.setSpan(
                            ForegroundColorSpan(theme.tempColor), textEnd, tempEnd,
                            SpannableString.SPAN_INCLUSIVE_INCLUSIVE
                        )
                        ss.setSpan(
                            ForegroundColorSpan(theme.humidColor), tempEnd, humidEnd,
                            SpannableString.SPAN_INCLUSIVE_INCLUSIVE
                        )
                        ss.setSpan(
                            ForegroundColorSpan(theme.rainColor), humidEnd, rainEnd,
                            SpannableString.SPAN_INCLUSIVE_INCLUSIVE
                        )
                        views.setTextViewText(cols[col], ss)

                        col++
                    }
                }
            }
            return views
        }

        var mIconManager: WeatherIconManager? = null

        fun getIconManager(context: Context): WeatherIconManager {
            if (mIconManager == null) {
                mIconManager = WeatherIconManager(context)
            }
            return mIconManager!!
        }
    }


    internal class WeatherIconManager(private val mContext: Context) {
        val fetcher = WidgetUpdateServiceComponent().fetcher
        private val mCache = HashMap<String, Bitmap>()
        var mIconSize: Int
        var mShadowOffset: Int

        init {
            val density = mContext.resources.displayMetrics.density
            mShadowOffset = Math.round(1 * density)
            mIconSize = (38 * density).toInt() + mShadowOffset * 2
        }

        fun getIcon(url: String): Bitmap? {
            try {
                var bmp = mCache[url]
                if (bmp != null) return bmp

                bmp = runBlocking { fetcher.getImage2(url) }
                if (bmp != null) {
                    val newBitmap = convertBitmap(bmp)
                    if (newBitmap != null) {
                        bmp = newBitmap
                    }
                    mCache[url] = bmp
                }
                return bmp
            } catch (e: Exception) {
                e.printStackTrace()
                return null
            }
        }

        fun convertBitmap(bmp: Bitmap): Bitmap {
            val newBmp = Bitmap.createBitmap(mIconSize, mIconSize, Bitmap.Config.ARGB_8888)
            val alpha = bmp.extractAlpha()
            try {
                val c = Canvas(newBmp)
                val p = Paint()
                p.isFilterBitmap = true
                p.color = -0x78000000
                val m = Matrix()
                val scaleX = (mIconSize - mShadowOffset * 2).toFloat() / bmp.width
                val scaleY = (mIconSize - mShadowOffset * 2).toFloat() / bmp.height
                val scale = max(scaleX.toDouble(), scaleY.toDouble()).toFloat()
                m.preScale(scale, scale)
                m.postTranslate((mShadowOffset * 2).toFloat(), (mShadowOffset * 2).toFloat())
                c.drawBitmap(alpha, m, p)
                m.postTranslate(-mShadowOffset.toFloat(), -mShadowOffset.toFloat())
                c.drawBitmap(bmp, m, null)
            } finally {
                alpha.recycle()
            }

            return newBmp
        }
    }

    class WidgetTheme(context: Context?) {
        //			Resources res = context.getResources();
        val timeColor: Int = -0x2f2f30
        val tempColor: Int = -0x53a7
        val rainColor: Int = -0x713801
        val humidColor: Int = -0x710072
        val textColor: Int = -0x1f1f20
    }

    companion object {
        val TAG: String = WidgetUpdateService::class.java.simpleName
        const val ACTION_MANUAL_UPDATE: String = "manualUpdate"

        private val bmpNames = arrayOf(
            "psun", "psnow", "psleet",
            "prain_light", "prain_gusty",
            "prain", "pmoon", "pclouds"
        )

        private val bmpIds = intArrayOf(
            R.drawable.psun, R.drawable.psnow, R.drawable.psleet,
            R.drawable.prain, R.drawable.prain_gusty,
            R.drawable.prain, R.drawable.pmoon, R.drawable.pclouds
        )

        fun getBitmapIndexFromUrl(c: Context?, url: String?): Int {
            return getBitmapIndexFromUrl_old(c, url)
        }

        fun getBitmapIndexFromUrl_old(c: Context?, url: String?): Int {
            if (url != null) {
                for (j in bmpNames.indices) {
                    if (url.contains(bmpNames[j])) {
                        return bmpIds[j]
                    }
                }
            }
            return -1
        }
    }
}
